// @vaadin/hilla-file-router/runtime/createMenuItems.js
function __REGISTER__(feature, vaadinObj = window.Vaadin ??= {}) {
  vaadinObj.registrations ??= [];
  vaadinObj.registrations.push({
    is: feature ? `${"@vaadin/hilla-file-router"}/${feature}` : "@vaadin/hilla-file-router",
    version: "24.4.3"
  });
}
import { RouteParamType } from "../shared/routeParamType.js";
function createMenuItems(vaadinObject = window.Vaadin) {
  __REGISTER__("createMenuItems", vaadinObject);
  const collator = new Intl.Collator("en-US");
  return vaadinObject?.views ? Object.entries(vaadinObject.views).filter(
    ([_key, value]) => !value.menu?.exclude && !(value.params && Object.values(value.params).some((p) => p === RouteParamType.Required))
  ).map(([path, config]) => {
    const _path = config.params ? Object.keys(config.params).reduce((acc, key) => acc.replaceAll(key, ""), path) : path;
    return {
      to: _path,
      icon: config.menu?.icon,
      title: config.menu?.title ?? config.title,
      order: config.menu?.order
    };
  }).sort((menuA, menuB) => {
    const ordersDiff = (menuA.order ?? Number.MAX_VALUE) - (menuB.order ?? Number.MAX_VALUE);
    return ordersDiff !== 0 ? ordersDiff : collator.compare(menuA.to, menuB.to);
  }) : [];
}
export {
  createMenuItems
};
